docs/design/kademlia-removal.md

```diff
@@ -0,0 +1,84 @@
+# Kademlia Removal
+
+## Abstract
+
+This design document outlines the communication protocol between satellites and 
+storage nodes, network refreshes, and kademlia removal given the satellite opt-in 
+capability for storage nodes. 
+
+## Background
+User research on how Storage Node Operators would connect to satellites resulted in a 
+simple solution called Opt-in SNO select. Satellites wait for storage nodes to work 
+with them, and storage nodes manually select the satellites they want to work with. 
+The initial implementation of this allows SNOs to update their trusted satellite list 
+in a hardcoded configuration file, but future improvements will enable users to manage 
+this list through a web console. Due to the nature of this change, the Storj network 
+at this point in time no longer requires the Kademlia DHT or related entities. 
+We will replace this protocol with direct communication between satellites and 
+storage nodes, and keep the network fresh without kademlia node discovery and random 
+lookups.
+
+## Design
+
+1. Nodes reach out to satellites that they want to work with
+    
+    - At the time of writing, a storage node operator can input a list of trusted satellite 
+    IDs and addresses into their configuration file on setup. There are several tardigrade-level 
+    satellites already listed as a preset. This list is imported into a Go map data structure 
+    within the “trust” package.
+    - Eventually during the design and implementation of trusted satellite management, 
+    this configured list may be imported into a on-disk data store, eg SQL table. Users will 
+    be able to modify this list through the storage node operator console. However, this is a 
+    separate discussion from this design document. This design is satisfied with a static trusted 
+    satellites list.
+    - Nodes will be able to communicate with satellites through the “transport” and “trust” packages 
+    directly rather than using kademlia to traverse the network to find the address of a given ID.
+
+2. Network refreshes at a regular interval
+    - Nodes will keep themselves up to date in the network by pinging all the satellites in their 
+    trusted list every hour. 
+    - Satellites will ping the nodes back
+        - If is it successful, the satellite will insert or update the node in the overlay cache and 
+        notify the node of success
+        - If not, it does not proceed with updating the overlay cache, and the node receives an error message
+    - This will use grpc for now, but future development may use a non-ssl connection from node to satellite 
+    to initiate the refresh connection
+    - Update the overlay cache refresh method to communicate with nodes directly, without relying on 
+    kademlia lookups. Iterate through cache and dial each node.
```

Comment:
I'm still skeptical that we need the server to reach out to the storage nodes.  I feel at minimum we need a sentence that justifies keeping this functionality.

Reply:
I think the rationale was that if the satellite only relies on the nodes pinging _it_, then a node could wake up once every hour to contact the satellite, then go offline. However, they're going to fail audits if they do that, so I'm not sure that's a viable strategy anyway.

There must be a reason that I'm forgetting; Couldn't the cache be updated when SNs contact the satellite every hour?

Reply:
Clarifying, I understand we need to make sure the node is reachable by the server.  But, I feel like "Update the overlay cache refresh method" implies that this is happening on an interval and initiated by the server.  Still not sure about that part.  Maybe a process which reaches out to SNs not contacted for other reasons in the last hour.  That would add some randomness to the timings.

Reply:
okay, this is the reaper part - we should iterate through and only dial nodes we haven't heard from recently (in the last hour)

Reply:
I understand this except for the "dial nodes" part. If they haven't pinged us, why should we ping them back? The node is violating the protocol at that point, and perhaps its reputation could suffer, but I don't see a reason to dial it explicitly ... ?

Or rather, I only really see a need for one or the other. If the node is in charge of pinging the satellite every hour, the node is responsible for keeping its "uptime" good. Or, if the satellite is in charge of pinging the node periodically, the satellite is in charge of knowing each node's uptime. But I don't see the need for both the node to ping the satellite and the satellite to ping offline nodes.

Reply:
it may be the case that for whatever reason, the satellite missed a bunch of udp packets. we don't want to consider storage nodes offline unless they're really offline, especially if they have lots of data that we care about.

Reply:
in my view, pinging nodes we haven't heard from is a robustness thing. less single points/systems of failure

Reply:
In that case, could we not simply use TCP (without TLS) and then know that we haven't missed any udp packets? I see your point, but I do worry that the small win we get from using UDP creates other complications that may not seem worth it. (Remember our place of former employment, and how they used udp so much, and how it was a complete nightmare ... ? :smile_cat:)

Reply:
udp is great, maaaan!

okay, let's say we know for sure storage nodes will continue to send us packets until we get them (what switching to unauthenticated TCP would essentially provide). then the satellite goes offline for 2 hours. it's still more robust to make sure the satellite doesn't penalize anyone until it's double checked, right?

Reply:
Yes, but there are other ways to be sure you aren't prematurely penalizing. For example, if the satellite knows its policy is that nodes check in every hour, when it comes online, it knows to wait at least that long before penalizing nodes. Seems simple.

I love UDP as much as the next guy, but sometimes delivery guarantees are pretty nice, ya know? For server and client.

Reply:
but what if the satellite doesn't know it went offline (network connection link dipped for a bit, etc)? the ping double checking isn't only because of delivery guarantees. i agree with you that we could require that nodes check in with satellites once an hour and we penalize any node that we don't hear from. the udp part is a red herring - if we use UDP, a node gets a delivery guarantee if the satellite pings it back (okay, it got the udp packet). the question is, do we want to double check before we penalize?

Reply:
ok, I'm convinced. Dialing before penalizing is good, you're right it could have network difficulties it doesn't know about.

UDP is fine. I just feel like, in general, when using UDP for packets that you actual do care about, it can lead to long-term issues that are hard to anticipate now and we end up engineering around for years to come. But I'll admit I can't foresee them now, I just have warning bells that go off when we use udp for packets are, actually, important.

Reply:
well the Achilles heel in my argument i don't know what to do about is... what if the network interruption is still ongoing when we try to ping back? what if the satellite was disconnected from the internet via tractor? what if only cogent/level3/comcast links were cut off via tractor? i don't think we've figured out a general solution to that problem yet

Reply:
https://www.theguardian.com/technology/2014/aug/14/google-undersea-fibre-optic-cables-shark-attacks